### DISCUSSION TITLE
First-party support for `<picture>` with `astro:assets`

### DISCUSSION THREAD

**Discussion by tony-sull** [withastro/roadmap#598] [Proposal]
### Body

# Summary

`astro:assets` is a huge leap forward adding image optimization support directly in Astro core 🎉  Now that it's out the door, in no small part by keeping the scope limited to `<Image>`, I'd be interested to see support building multiple image formats for the [`<picture>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture)

# Background & Motivation

Images on the web are tricky 😅

When it comes to performance, it's best for pages to load the correct size *and* image format based on the browser's support and window size.  Multiple sizes can be handled by `<img srcset="">`, but only the `<picture>` element has support for defining multiple image formats like `.jpg`, `.webp`, and `.avif` for the browser to select from.

## Hosted services

Some hosted image services get around this limitation by sniffing request headers to determine the best image format supported by each browser requesting the image.  [Cloudinary](https://cloudinary.com/) and [Vercel](https://vercel.com/docs/concepts/image-optimization) are both examples of this, allowing sites to request the original image by URL and leaving it up to the hosted service to decide what image format to use.

This solution works well for server-rendered apps, and this could potentially be a solution for SSR apps with `astro:assets`, which injects an API endpoint for transforming images when requested instead of during the build. This wouldn't work for SSG projects though, which need to generate all image formats statically during the build.

# Goals

- include a `<Picture>` component in `astro:assets` that supports building multiple and formats **and** sizes

## Out of scope

supporting the [art direction](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#art_direction) use case where different sizes also have different aspect ratios
- **Why?** This adds quite a bit of complexity to the API requirements
- any APIs needed to handle this in a third-party component are available even thought `astro:assets` doesn't provide this out of the box

adding header sniffing support into the injected `_image` API route used to optimize images in SSR pages
- **Why?** Any solution that works for multiple formats in SSG builds would also handle those use cases in SSR without header sniffing or similar

# Example

`@astrojs/image` includes a `<Picture>` component that may be a decent starting point.  It doesn't cover art direction but does include support for users providing a list of multiple widths and formats to be used.

```astro
---
import { Picture } from "@astrojs/image/components"
---

<Picture widths={[450, 900, 1800]} formats={["avif", "webp", "jpg"]} />
```

This may be a good starting point if supporting multiple widths & formats covers the needs for most Astro projects.  This API would need a redesign if art direction is more important to the community.

**Comment by louiss0** [comment]
I completely agree with this I hope this is a thing in Astro. Having to write source attributes is too painful.

**Comment by cquesadad** [comment]
I think it is needed in order to make native the solution and not depending on other external packages. Thanks

**Comment by AndreCox** [comment]
This is 100% needed I can't find a convienient way of doing this right now.

**Reply by Princesseuh** [reply]
You can achieve this right now using multiple calls to `getImage`. A core implementation would really just be a wrapper around this.

**Comment by jasikpark** [comment]
> This may be a good starting point if supporting multiple widths & formats covers the needs for most Astro projects. This API would need a redesign if art direction is more important to the community.

IMO using Picture w/ art direction should follow the same API as the native <picture> element, with nested <Image>'s inside of it, which allows for the sort of control over art direction w/o a bespoke styling api

**Reply by louiss0** [reply]
I think it's better to think of this as more of a convenience feature. Manually writing multiple source tags is a pain. As a dev I do just care about specifying the widths.

**Reply by jasikpark** [reply]
Right, when you only care about specifiying the widths, the simplified layout @tony-sull suggested works perfectly, and then you could drop down to the composable layout when you really need the art direction...

**Comment by sannajammeh** [comment]
I am currently porting Next 13's Image functionality to Astro. Here is the WIP gist: https://gist.github.com/sannajammeh/d1dfddc8c2538062de53714b07953cdf

Currently working:
- [x] Astro assets compatibility
- [x] Responsive sizes generation (using native browser srcset and sizes)
- [x] Vercel's imageService - skip long build times and build images on demand
- [x] Configurable sizes prop
- [x] `deviceWidth` prop to override default [380, 640, ... etc] - if using Vercel, set also in integration config
- [ ] `priority` flag - If anyone knows how to inject link tags into `head` through Astro's rendering process please let me know.
- [x] `placeholder="blur"` - Working (Vercel, standalone)

**Reply by natemoo-re** [reply]
`priority` flag could definitely be implemented on the Astro side! Don't think it's possible in userland right now, but we have an internal rendering API (`createHeadAndContent`) that can pass tags up to the `head` before rendering a component.

**Reply by sannajammeh** [reply]
> `priority` flag could definitely be implemented on the Astro side! Don't think it's possible in userland right now, but we have an internal rendering API (`createHeadAndContent`) that can pass tags up to the `head` before rendering a component.

I did some random attempts using object references to increment numbers nested within the tree and it did in fact log the incremented value in the root layout. Very unsure about how reliable that is though, but if it is reliable then I can implement something using `unhead` for SSR until a userland API lands.

**Reply by florian-lefebvre** [reply]
If you want an example of usage of unhead with astro, you can check out https://github.com/florian-lefebvre/astro-capo

**Reply by sannajammeh** [reply]
Now updated with `placeholder="blur"` support on Node & Vercel, probably others too.

**Comment by Princesseuh** [comment]
This proposal has been accepted! Please continue discussion in the issue: https://github.com/withastro/roadmap/issues/700

We'll be working on a Stage 3 proposal and an implementation this next quarter.
